entrypoint.d/45-geoip

```diff
@@ -0,0 +1,29 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+import re
+import sys
+
+from doodbalib import logger
+
+from plumbum import NOHUP, local
+from plumbum.cmd import geoipupdate
+
+GEOIP_ACCOUNT_ID = local.env["GEOIP_ACCOUNT_ID"]
+GEOIP_LICENSE_KEY = local.env["GEOIP_LICENSE_KEY"]
+
+if GEOIP_ACCOUNT_ID not in {"", "0"} and GEOIP_LICENSE_KEY not in {"", "0"}:
+    # see https://dev.maxmind.com/geoip/geoipupdate/
+    logger.info("Activating GeoIP/GeoLite2 updates")
+    logger.info(
+        "This product includes GeoLite2 data created by MaxMind, "
+        "available from https://www.maxmind.com"
+    )
+    # Obtain current GeoIP configuration
+    geoip_file = local.path("/etc/GeoIP.conf")
+    geoip_conf = geoip_file.read()
+    # Patch it with provided account and license keys
+    geoip_conf = re.sub(r"^AccountID\s*.*$", "AccountID {}".format(GEOIP_ACCOUNT_ID))
+    geoip_conf = re.sub(r"^LicenseKey\s*.*$", "LicenseKey {}".format(GEOIP_LICENSE_KEY))
+    geoip_file.write(geoip_conf)
+    # run geoip update in background to make sure it doesn't block odoo startup
+    geoipupdate & NOHUP(stdout=sys.stdout, stderr=sys.stderr)
```

Comment:
Sorry to bother you again my friend.

Could you please make sure again this works as expected and detaches the geoipupdate process, lets it continue even when the entrypoint script is gone, and is not left as zombie if odoo stops before geoip ends updating?

I just want to make sure this works as expected :sweat_smile: 

BTW you see here the same script in python + plumbum. :sparkles: :heart_eyes: :sparkles:

Reply:
I fixed the missing argument for re.sub in https://github.com/Tecnativa/doodba/pull/275/commits/8518587aa41a6c7f9d3738800b3395bd24d630b0 but i still get errors:
```
doodba INFO: Waiting until postgres is listening at db...
doodba INFO: Unaccent extension installed
doodba INFO: Linking all addons from /opt/odoo/custom/src/addons.yaml in /opt/odoo/auto/addons
doodba INFO: Activating GeoIP/GeoLite2 updates
doodba INFO: This product includes GeoLite2 data created by MaxMind, available from https://www.maxmind.com
Traceback (most recent call last):
  File "/opt/odoo/common/entrypoint.d/45-geoip", line 29, in <module>
    geoipupdate & NOHUP(stdout=sys.stdout, stderr=sys.stderr)
  File "/usr/local/lib/python3.7/site-packages/plumbum/commands/modifiers.py", line 377, in __rand__
    return cmd.nohup(cmd, self.cwd, stdout, self.stderr, append)
  File "/usr/local/lib/python3.7/site-packages/plumbum/commands/base.py", line 152, in nohup
    return self.machine.daemonic_popen(self, cwd, stdout, stderr, append)
  File "/usr/local/lib/python3.7/site-packages/plumbum/machines/local.py", line 337, in daemonic_popen
    return posix_daemonize(command, cwd, stdout, stderr, append)
  File "/usr/local/lib/python3.7/site-packages/plumbum/commands/daemons.py", line 70, in posix_daemonize
    raise ProcessExecutionError(argv, rc, "", output)
plumbum.commands.processes.ProcessExecutionError: Unexpected exit code: 256
Command line: | /usr/bin/geoipupdate
Stderr:       | Traceback (most recent call last):
              |   File "/usr/local/lib/python3.7/site-packages/plumbum/commands/daemons.py", line 39, in posix_daemonize
              |     stdout = open(stdout, "a" if append else "w")
              | TypeError: expected str, bytes or os.PathLike object, not _io.TextIOWrapper
Traceback (most recent call last):
  File "/opt/odoo/common/entrypoint", line 32, in <module>
    subprocess.check_call(command)
  File "/usr/local/lib/python3.7/subprocess.py", line 347, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '/opt/odoo/common/entrypoint.d/45-geoip' returned non-zero exit status 1.
```

does this mean exit code 256 from geoipupdate or does this mean something else?

Reply:
also fixed the regexes now (still getting the same error)

Reply:
ok i give up, i cannot find NOHUP in the plumbum documentation
the problem seems to be the `geoipupdate & NOHUP(stdout=sys.stdout, stderr=sys.stderr)`
i tested on a python console with the following:
```
bigbear3001@wt-io-it-bigbear3001:~$ python3 -c 'from plumbum import NOHUP, BG; from plumbum.cmd import bash; import sys; bash["-c", "sleep 1; echo 'finished'"] & NOHUP(stdout=sys.stdout)'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/bigbear3001/.local/lib/python3.6/site-packages/plumbum/commands/modifiers.py", line 377, in __rand__
    return cmd.nohup(cmd, self.cwd, stdout, self.stderr, append)
  File "/home/bigbear3001/.local/lib/python3.6/site-packages/plumbum/commands/base.py", line 152, in nohup
    return self.machine.daemonic_popen(self, cwd, stdout, stderr, append)
  File "/home/bigbear3001/.local/lib/python3.6/site-packages/plumbum/machines/local.py", line 337, in daemonic_popen
    return posix_daemonize(command, cwd, stdout, stderr, append)
  File "/home/bigbear3001/.local/lib/python3.6/site-packages/plumbum/commands/daemons.py", line 70, in posix_daemonize
    raise ProcessExecutionError(argv, rc, "", output)
plumbum.commands.processes.ProcessExecutionError: Unexpected exit code: 256
Command line: | /bin/bash -c 'sleep 1; echo finished'
Stderr:       | Traceback (most recent call last):
              |   File "/home/bigbear3001/.local/lib/python3.6/site-packages/plumbum/commands/daemons.py", line 39, in posix_daemonize
              |     stdout = open(stdout, "a" if append else "w")
              | TypeError: expected str, bytes or os.PathLike object, not _io.TextIOWrapper
```
but BG seems to do the trick:
```
bigbear3001@wt-io-it-bigbear3001:~$ python3 -c 'from plumbum import NOHUP, BG; from plumbum.cmd import bash; import sys; bash["-c", "sleep 1; echo 'finished'"] & BG(stdout=sys.stdout)'
bigbear3001@wt-io-it-bigbear3001:~$ finished
```

Reply:
Retested after the latest commit: https://github.com/Tecnativa/doodba/pull/275/commits/495c6bfc5924dfcd58d43e3daca6b9e6e53ec9f9

* geoipupdate does not block odoo startup
* geoipupdate terminates as soon as odoo terminates when running `docker-compose up` and pressing CTRL+C
* geoipupdate error message is shown in `docker-compose logs` of odoo container (it does not fail silently)
* after starting maxmind_proxy odoo needs a restart / workers need to restart to activate geoip functionality

Reply:
Excellent!

NOHUP is documented here: https://plumbum.readthedocs.io/en/latest/local_commands.html#background-and-foreground

According to the docs, I thought that using BG would stop the background job as soon as the main job exited, but it seems to not do that, so... perfect!

Thanks a lot.